In [53]:
import numpy as np
import pandas as pd
import json

<h1>
<font color="#15aec3"> Pandas: Limpeza e tratamento de dados</font>
</h1>

Obtenção da base:

```bash
! curl https://cdn3.gnarususercontent.com.br/2929-pandas/dataset-telecon.json > telecon.json
```

<h2>
<font color="#9de482"> Análise de Churn </font>
</h2>

<h2>
<font color="#9de482" > 01 - Conhecendo os dados</font>
</h2>

In [54]:
telecon_path: str = "./data/telecon.json"

df = pd.read_json(telecon_path)

In [55]:
df

id_cliente Churn                                            cliente  \
0     0002-ORFBO   nao  {'genero': 'feminino', 'idoso': 0, 'parceiro':...   
1     0003-MKNFE   nao  {'genero': 'masculino', 'idoso': 0, 'parceiro'...   
2     0004-TLHLJ   sim  {'genero': 'masculino', 'idoso': 0, 'parceiro'...   
3     0011-IGKFF   sim  {'genero': 'masculino', 'idoso': 1, 'parceiro'...   
4     0013-EXCHZ   sim  {'genero': 'feminino', 'idoso': 1, 'parceiro':...   
...          ...   ...                                                ...   
7339  5172-RKOCB   nao  {'genero': 'masculino', 'idoso': 0, 'parceiro'...   
7340  1934-MKPXS   nao  {'genero': 'masculino', 'idoso': 0, 'parceiro'...   
7341  5959-BELXA   sim  {'genero': 'masculino', 'idoso': 1, 'parceiro'...   
7342  3601-UTZXO        {'genero': 'masculino', 'idoso': 0, 'parceiro'...   
7343  4452-QIIEB   nao  {'genero': 'masculino', 'idoso': 0, 'parceiro'...   

                                               telefone  \
0     {'servico_telefone': 'sim', 'varias_linhas': '...   
1     {'servico_telefone': 'sim', 'varias_linhas': '...   
2     {'servico_telefone': 'sim', 'varias_linhas': '...   
3     {'servico_telefone': 'sim', 'varias_linhas': '...   
4     {'servico_telefone': 'sim', 'varias_linhas': '...   
...                                                 ...   
7339  {'servico_telefone': 'sim', 'varias_linhas': '...   
7340  {'servico_telefone': 'sim', 'varias_linhas': '...   
7341  {'servico_telefone': 'sim', 'varias_linhas': '...   
7342  {'servico_telefone': 'sim', 'varias_linhas': '...   
7343  {'servico_telefone': 'sim', 'varias_linhas': '...   

                                               internet  \
0     {'servico_internet': 'DSL', 'seguranca_online'...   
1     {'servico_internet': 'DSL', 'seguranca_online'...   
2     {'servico_internet': 'fibra otica', 'seguranca...   
3     {'servico_internet': 'fibra otica', 'seguranca...   
4     {'servico_internet': 'fibra otica', 'seguranca...   
...                                                 ...   
7339  {'servico_internet': 'fibra otica', 'seguranca...   
7340  {'servico_internet': 'nao', 'seguranca_online'...   
7341  {'servico_internet': 'fibra otica', 'seguranca...   
7342  {'servico_internet': 'nao', 'seguranca_online'...   
7343  {'servico_internet': 'DSL', 'seguranca_online'...   

                                                  conta  
0     {'contrato': None, 'faturamente_eletronico': N...  
1     {'contrato': 'mes a mes', 'faturamente_eletron...  
2     {'contrato': 'mes a mes', 'faturamente_eletron...  
3     {'contrato': 'mes a mes', 'faturamente_eletron...  
4     {'contrato': 'mes a mes', 'faturamente_eletron...  
...                                                 ...  
7339  {'contrato': 'dois anos', 'faturamente_eletron...  
7340  {'contrato': 'um ano', 'faturamente_eletronico...  
7341  {'contrato': 'mes a mes', 'faturamente_eletron...  
7342  {'contrato': 'um ano', 'faturamente_eletronico...  
7343  {'contrato': 'mes a mes', 'faturamente_eletron...  

[7344 rows x 6 columns]

In [56]:
pd.json_normalize(df['cliente'])

genero  idoso parceiro dependentes  tempo_servico
0      feminino      0      sim         sim            9.0
1     masculino      0      nao         nao            9.0
2     masculino      0      nao         nao            4.0
3     masculino      1      sim         nao           13.0
4      feminino      1      sim         nao            3.0
...         ...    ...      ...         ...            ...
7339  masculino      0      sim         nao           72.0
7340  masculino      0      sim         sim           33.0
7341  masculino      1      nao         nao           32.0
7342  masculino      0      sim         sim           41.0
7343  masculino      0      nao         nao            7.0

[7344 rows x 5 columns]

In [57]:
pd.json_normalize(df['conta'])

contrato faturamente_eletronico                     metodo_pagamento  \
0          None                   None                                 None   
1     mes a mes                    nao                  cheque pelo correio   
2     mes a mes                    sim                    cheque eletronico   
3     mes a mes                    sim                    cheque eletronico   
4     mes a mes                    sim                  cheque pelo correio   
...         ...                    ...                                  ...   
7339  dois anos                    sim       cartao de credito (automatico)   
7340     um ano                    nao       cartao de credito (automatico)   
7341  mes a mes                    sim       cartao de credito (automatico)   
7342     um ano                    nao  transferencia bancaria (automatica)   
7343  mes a mes                    nao                  cheque pelo correio   

      cobranca.mensal cobranca.Total  
0                 NaN           None  
1               59.90          542.4  
2               73.90         280.85  
3               98.00        1237.85  
4               83.90          267.4  
...               ...            ...  
7339           108.95           7875  
7340            20.10         620.55  
7341            96.15        3019.25  
7342            19.50          798.2  
7343            43.90          278.4  

[7344 rows x 5 columns]

Vale a pena usar a biblioteca padrão `json` para normalizar todas as colunas de uma vez.

In [58]:
with open(file=telecon_path) as file:
    raw_json = json.load(file)
    

In [59]:
dados = pd.json_normalize(raw_json)
dados

id_cliente Churn cliente.genero  cliente.idoso cliente.parceiro  \
0     0002-ORFBO   nao       feminino              0              sim   
1     0003-MKNFE   nao      masculino              0              nao   
2     0004-TLHLJ   sim      masculino              0              nao   
3     0011-IGKFF   sim      masculino              1              sim   
4     0013-EXCHZ   sim       feminino              1              sim   
...          ...   ...            ...            ...              ...   
7339  5172-RKOCB   nao      masculino              0              sim   
7340  1934-MKPXS   nao      masculino              0              sim   
7341  5959-BELXA   sim      masculino              1              nao   
7342  3601-UTZXO            masculino              0              sim   
7343  4452-QIIEB   nao      masculino              0              nao   

     cliente.dependentes  cliente.tempo_servico telefone.servico_telefone  \
0                    sim                    9.0                       sim   
1                    nao                    9.0                       sim   
2                    nao                    4.0                       sim   
3                    nao                   13.0                       sim   
4                    nao                    3.0                       sim   
...                  ...                    ...                       ...   
7339                 nao                   72.0                       sim   
7340                 sim                   33.0                       sim   
7341                 nao                   32.0                       sim   
7342                 sim                   41.0                       sim   
7343                 nao                    7.0                       sim   

     telefone.varias_linhas internet.servico_internet  ...  \
0                       nao                       DSL  ...   
1                       sim                       DSL  ...   
2                       nao               fibra otica  ...   
3                       nao               fibra otica  ...   
4                       nao               fibra otica  ...   
...                     ...                       ...  ...   
7339                    sim               fibra otica  ...   
7340                    nao                       nao  ...   
7341                    sim               fibra otica  ...   
7342                    nao                       nao  ...   
7343                    nao                       DSL  ...   

       internet.backup_online internet.protecao_dispositivo  \
0                         sim                           nao   
1                         nao                           nao   
2                         nao                           sim   
3                         sim                           sim   
4                         nao                           nao   
...                       ...                           ...   
7339                      sim                           nao   
7340  sem servico de internet       sem servico de internet   
7341                      nao                           nao   
7342  sem servico de internet       sem servico de internet   
7343                      nao                           nao   

     internet.suporte_tecnico    internet.tv_streaming  \
0                         sim                      sim   
1                         nao                      nao   
2                         nao                      nao   
3                         nao                      sim   
4                         sim                      sim   
...                       ...                      ...   
7339                      sim                      sim   
7340  sem servico de internet  sem servico de internet   
7341                      nao                      sim   
7342  sem servico de internet  sem servico de internet   
7343                      nao                      nao   

     

In [60]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7344 entries, 0 to 7343
Data columns (total 21 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   id_cliente                     7344 non-null   object 
 1   Churn                          7344 non-null   object 
 2   cliente.genero                 7344 non-null   object 
 3   cliente.idoso                  7344 non-null   int64  
 4   cliente.parceiro               7344 non-null   object 
 5   cliente.dependentes            7344 non-null   object 
 6   cliente.tempo_servico          7336 non-null   float64
 7   telefone.servico_telefone      7344 non-null   object 
 8   telefone.varias_linhas         7344 non-null   object 
 9   internet.servico_internet      7344 non-null   object 
 10  internet.seguranca_online      7344 non-null   object 
 11  internet.backup_online         7344 non-null   object 
 12  internet.protecao_dispositivo  7344 non-null   o

<h2>
<font color="#9de482"> 02 - Tratamento Inicial dos dados </font>
</h2>

In [61]:
dados.head()

id_cliente Churn cliente.genero  cliente.idoso cliente.parceiro  \
0  0002-ORFBO   nao       feminino              0              sim   
1  0003-MKNFE   nao      masculino              0              nao   
2  0004-TLHLJ   sim      masculino              0              nao   
3  0011-IGKFF   sim      masculino              1              sim   
4  0013-EXCHZ   sim       feminino              1              sim   

  cliente.dependentes  cliente.tempo_servico telefone.servico_telefone  \
0                 sim                    9.0                       sim   
1                 nao                    9.0                       sim   
2                 nao                    4.0                       sim   
3                 nao                   13.0                       sim   
4                 nao                    3.0                       sim   

  telefone.varias_linhas internet.servico_internet  ...  \
0                    nao                       DSL  ...   
1                    sim                       DSL  ...   
2                    nao               fibra otica  ...   
3                    nao               fibra otica  ...   
4                    nao               fibra otica  ...   

  internet.backup_online internet.protecao_dispositivo  \
0                    sim                           nao   
1                    nao                           nao   
2                    nao                           sim   
3                    sim                           sim   
4                    nao                           nao   

  internet.suporte_tecnico internet.tv_streaming internet.filmes_streaming  \
0                      sim                   sim                       nao   
1                      nao                   nao                       sim   
2                      nao                   nao                       nao   
3                      nao                   sim                       sim   
4                      sim                   sim                       nao   

  conta.contrato conta.faturamente_eletronico conta.metodo_pagamento  \
0           None                         None                   None   
1      mes a mes                          nao    cheque pelo correio   
2      mes a mes                          sim      cheque eletronico   
3      mes a mes                          sim      cheque eletronico   
4      mes a mes                          sim    cheque pelo correio   

  conta.cobranca.mensal  conta.cobranca.Total  
0                   NaN                  None  
1                  59.9                 542.4  
2                  73.9                280.85  
3                  98.0               1237.85  
4                  83.9                 267.4  

[5 rows x 21 columns]

In [62]:
dados.dtypes

id_cliente                        object
Churn                             object
cliente.genero                    object
cliente.idoso                      int64
cliente.parceiro                  object
cliente.dependentes               object
cliente.tempo_servico            float64
telefone.servico_telefone         object
telefone.varias_linhas            object
internet.servico_internet         object
internet.seguranca_online         object
internet.backup_online            object
internet.protecao_dispositivo     object
internet.suporte_tecnico          object
internet.tv_streaming             object
internet.filmes_streaming         object
conta.contrato                    object
conta.faturamente_eletronico      object
conta.metodo_pagamento            object
conta.cobranca.mensal            float64
conta.cobranca.Total              object
dtype: object

In [63]:
try:
    dados['conta.cobranca.Total'].astype(np.float64)
except ValueError as e:
    print(f"Erro ao converter coluna:")
    print(e)
    

Erro ao converter coluna:
could not convert string to float: ' '


In [64]:
# Colunas que deram erro no casting
dados[
    dados['conta.cobranca.Total'] == ' '
]

id_cliente Churn cliente.genero  cliente.idoso cliente.parceiro  \
975   1371-DWPAZ   nao       feminino              0              sim   
1775  2520-SGTTA   nao       feminino              0              sim   
1955  2775-SEFEE   nao      masculino              0              nao   
2075  2923-ARZLG   nao      masculino              0              sim   
2232  3115-CZMZD   nao      masculino              0              nao   
2308  3213-VVOLG   nao      masculino              0              sim   
2930  4075-WKNIU   nao       feminino              0              sim   
3134  4367-NUYAO   nao      masculino              0              sim   
3203  4472-LVYGI   nao       feminino              0              sim   
4169  5709-LVOEQ   nao       feminino              0              sim   
5599  7644-OMVMY   nao      masculino              0              sim   

     cliente.dependentes  cliente.tempo_servico telefone.servico_telefone  \
975                  sim                    0.0                       nao   
1775                 sim                    0.0                       sim   
1955                 sim                    0.0                       sim   
2075                 sim                    0.0                       sim   
2232                 sim                    0.0                       sim   
2308                 sim                    0.0                       sim   
2930                 sim                    0.0                       sim   
3134                 sim                    0.0                       sim   
3203                 sim                    0.0                       nao   
4169                 sim                    0.0                       sim   
5599                 sim                    0.0                       sim   

       telefone.varias_linhas internet.servico_internet  ...  \
975   sem servico de telefone                       DSL  ...   
1775                      nao                       nao  ...   
1955                      sim                       DSL  ...   
2075                      nao                       nao  ...   
2232                      nao                       nao  ...   
2308                      sim                       nao  ...   
2930                      sim                       DSL  ...   
3134                      sim                       nao  ...   
3203  sem servico de telefone                       DSL  ...   
4169                      nao                       DSL  ...   
5599                      nao                       nao  ...   

       internet.backup_online internet.protecao_dispositivo  \
975                       sim                           sim   
1775  sem servico de internet       sem servico de internet   
1955                      sim                           nao   
2075  sem servico de internet       sem servico de internet   
2232  sem servico de internet       sem servico de internet   
2308  sem servico de internet       sem servico de internet   
2930                      sim                           sim   
3134  sem servico de internet       sem servico de internet   
3203                      nao                           sim   
4169                      sim                           sim   
5599  sem servico de internet       sem servico de internet   

     internet.suporte_tecnico    internet.tv_streaming  \
975                       sim                      sim   
1775  sem servico de internet  sem servico de internet   
1955                      sim                      nao   
2075  sem servico de internet  sem servico de internet   
2232  sem servico de internet  sem servico de internet   
2308  sem servico de internet  sem servico de internet   
2930                      sim                      sim   
3134  sem servico de internet  sem servico de internet   
3203                      sim                      sim   
4169                      nao                      sim   
5599  sem servico de internet  sem servi

In [65]:
registros_n_preenchidos = dados['conta.cobranca.Total'] == ' '
cols_interesse = ['cliente.tempo_servico', 'conta.contrato', 'conta.cobranca.mensal', 'conta.cobranca.Total']
dados[registros_n_preenchidos][cols_interesse]

cliente.tempo_servico conta.contrato  conta.cobranca.mensal  \
975                     0.0      dois anos                  56.05   
1775                    0.0      dois anos                  20.00   
1955                    0.0      dois anos                  61.90   
2075                    0.0      dois anos                  19.70   
2232                    0.0      dois anos                  20.25   
2308                    0.0      dois anos                  25.35   
2930                    0.0      dois anos                  73.35   
3134                    0.0      dois anos                  25.75   
3203                    0.0      dois anos                  52.55   
4169                    0.0      dois anos                  80.85   
5599                    0.0      dois anos                  19.85   

     conta.cobranca.Total  
975                        
1775                       
1955                       
2075                       
2232                       
2308                       
2930                       
3134                       
3203                       
4169                       
5599

In [66]:
indexes = dados[registros_n_preenchidos][cols_interesse].index

dados.loc[indexes, 'conta.cobranca.Total'] =dados.loc[indexes, 'conta.cobranca.mensal'] * 24 # dois anos

In [67]:
dados[registros_n_preenchidos][cols_interesse]

cliente.tempo_servico conta.contrato  conta.cobranca.mensal  \
975                     0.0      dois anos                  56.05   
1775                    0.0      dois anos                  20.00   
1955                    0.0      dois anos                  61.90   
2075                    0.0      dois anos                  19.70   
2232                    0.0      dois anos                  20.25   
2308                    0.0      dois anos                  25.35   
2930                    0.0      dois anos                  73.35   
3134                    0.0      dois anos                  25.75   
3203                    0.0      dois anos                  52.55   
4169                    0.0      dois anos                  80.85   
5599                    0.0      dois anos                  19.85   

     conta.cobranca.Total  
975                1345.2  
1775                480.0  
1955               1485.6  
2075                472.8  
2232                486.0  
2308                608.4  
2930               1760.4  
3134                618.0  
3203               1261.2  
4169               1940.4  
5599                476.4

In [68]:
dados.loc[indexes, "cliente.tempo_servico"] = 24

In [69]:
dados[registros_n_preenchidos][cols_interesse]

cliente.tempo_servico conta.contrato  conta.cobranca.mensal  \
975                    24.0      dois anos                  56.05   
1775                   24.0      dois anos                  20.00   
1955                   24.0      dois anos                  61.90   
2075                   24.0      dois anos                  19.70   
2232                   24.0      dois anos                  20.25   
2308                   24.0      dois anos                  25.35   
2930                   24.0      dois anos                  73.35   
3134                   24.0      dois anos                  25.75   
3203                   24.0      dois anos                  52.55   
4169                   24.0      dois anos                  80.85   
5599                   24.0      dois anos                  19.85   

     conta.cobranca.Total  
975                1345.2  
1775                480.0  
1955               1485.6  
2075                472.8  
2232                486.0  
2308                608.4  
2930               1760.4  
3134                618.0  
3203               1261.2  
4169               1940.4  
5599                476.4

In [70]:
dados['conta.cobranca.Total']  = dados['conta.cobranca.Total'].astype(np.float64)

In [71]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7344 entries, 0 to 7343
Data columns (total 21 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   id_cliente                     7344 non-null   object 
 1   Churn                          7344 non-null   object 
 2   cliente.genero                 7344 non-null   object 
 3   cliente.idoso                  7344 non-null   int64  
 4   cliente.parceiro               7344 non-null   object 
 5   cliente.dependentes            7344 non-null   object 
 6   cliente.tempo_servico          7336 non-null   float64
 7   telefone.servico_telefone      7344 non-null   object 
 8   telefone.varias_linhas         7344 non-null   object 
 9   internet.servico_internet      7344 non-null   object 
 10  internet.seguranca_online      7344 non-null   object 
 11  internet.backup_online         7344 non-null   object 
 12  internet.protecao_dispositivo  7344 non-null   o

Identificando e manipulando strings vazias

In [72]:
dados.query("Churn ==''")

id_cliente Churn cliente.genero  cliente.idoso cliente.parceiro  \
30    0047-ZHDTW             feminino              0              nao   
75    0120-YZLQA            masculino              0              nao   
96    0154-QYHJU            masculino              0              nao   
98    0162-RZGMZ             feminino              1              nao   
175   0274-VVQOQ            masculino              1              sim   
...          ...   ...            ...            ...              ...   
7211  9920-GNDMB            masculino              0              nao   
7239  9955-RVWSC             feminino              0              sim   
7247  9966-VYRTZ             feminino              0              sim   
7267  6532-YOHZY            masculino              0              sim   
7342  3601-UTZXO            masculino              0              sim   

     cliente.dependentes  cliente.tempo_servico telefone.servico_telefone  \
30                   nao                   11.0                       sim   
75                   nao                   71.0                       sim   
96                   nao                   29.0                       sim   
98                   nao                    5.0                       sim   
175                  nao                   65.0                       sim   
...                  ...                    ...                       ...   
7211                 nao                    9.0                       sim   
7239                 sim                   67.0                       sim   
7247                 sim                   31.0                       sim   
7267                 sim                   45.0                       sim   
7342                 sim                   41.0                       sim   

     telefone.varias_linhas internet.servico_internet  ...  \
30                      sim               fibra otica  ...   
75                      nao                       nao  ...   
96                      nao                       DSL  ...   
98                      nao                       DSL  ...   
175                     sim               fibra otica  ...   
...                     ...                       ...  ...   
7211                    sim               fibra otica  ...   
7239                    nao                       nao  ...   
7247                    nao                       nao  ...   
7267                    sim               fibra otica  ...   
7342                    nao                       nao  ...   

       internet.backup_online internet.protecao_dispositivo  \
30                        nao                           nao   
75    sem servico de internet       sem servico de internet   
96                        sim                           nao   
98                        sim                           nao   
175                       sim                           sim   
...                       ...                           ...   
7211                      nao                           nao   
7239  sem servico de internet       sem servico de internet   
7247  sem servico de internet       sem servico de internet   
7267                      sim                           sim   
7342  sem servico de internet       sem servico de internet   

     internet.suporte_tecnico    internet.tv_streaming  \
30                        nao                      nao   
75    sem servico de internet  sem servico de internet   
96                        sim                      nao   
98                        sim                      nao   
175                       nao                      sim   
...                       ...                      ...   
7211                      nao                      nao   
7239  sem servico de internet  sem servico de internet   
7247  sem servico de internet  sem servico de internet   
7267                      sim                      sim   
7342  sem servico de internet  sem servico de internet   

     

226 amostras em `Churn` estão vazias!

In [73]:
dados_sem_vazio = dados[
    dados['Churn'] != ''    
].copy()

In [74]:
dados_sem_vazio

id_cliente Churn cliente.genero  cliente.idoso cliente.parceiro  \
0     0002-ORFBO   nao       feminino              0              sim   
1     0003-MKNFE   nao      masculino              0              nao   
2     0004-TLHLJ   sim      masculino              0              nao   
3     0011-IGKFF   sim      masculino              1              sim   
4     0013-EXCHZ   sim       feminino              1              sim   
...          ...   ...            ...            ...              ...   
7338  5792-JALQC   nao       feminino              1              nao   
7339  5172-RKOCB   nao      masculino              0              sim   
7340  1934-MKPXS   nao      masculino              0              sim   
7341  5959-BELXA   sim      masculino              1              nao   
7343  4452-QIIEB   nao      masculino              0              nao   

     cliente.dependentes  cliente.tempo_servico telefone.servico_telefone  \
0                    sim                    9.0                       sim   
1                    nao                    9.0                       sim   
2                    nao                    4.0                       sim   
3                    nao                   13.0                       sim   
4                    nao                    3.0                       sim   
...                  ...                    ...                       ...   
7338                 nao                   52.0                       sim   
7339                 nao                   72.0                       sim   
7340                 sim                   33.0                       sim   
7341                 nao                   32.0                       sim   
7343                 nao                    7.0                       sim   

     telefone.varias_linhas internet.servico_internet  ...  \
0                       nao                       DSL  ...   
1                       sim                       DSL  ...   
2                       nao               fibra otica  ...   
3                       nao               fibra otica  ...   
4                       nao               fibra otica  ...   
...                     ...                       ...  ...   
7338                    sim                       DSL  ...   
7339                    sim               fibra otica  ...   
7340                    nao                       nao  ...   
7341                    sim               fibra otica  ...   
7343                    nao                       DSL  ...   

       internet.backup_online internet.protecao_dispositivo  \
0                         sim                           nao   
1                         nao                           nao   
2                         nao                           sim   
3                         sim                           sim   
4                         nao                           nao   
...                       ...                           ...   
7338                      nao                           sim   
7339                      sim                           nao   
7340  sem servico de internet       sem servico de internet   
7341                      nao                           nao   
7343                      nao                           nao   

     internet.suporte_tecnico    internet.tv_streaming  \
0                         sim                      sim   
1                         nao                      nao   
2                         nao                      nao   
3                         nao                      sim   
4                         sim                      sim   
...                       ...                      ...   
7338                      nao                      nao   
7339                      sim                      sim   
7340  sem servico de internet  sem servico de internet   
7341                      nao                      sim   
7343                      nao                      nao   

     

In [75]:
dados_sem_vazio.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7118 entries, 0 to 7343
Data columns (total 21 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   id_cliente                     7118 non-null   object 
 1   Churn                          7118 non-null   object 
 2   cliente.genero                 7118 non-null   object 
 3   cliente.idoso                  7118 non-null   int64  
 4   cliente.parceiro               7118 non-null   object 
 5   cliente.dependentes            7118 non-null   object 
 6   cliente.tempo_servico          7110 non-null   float64
 7   telefone.servico_telefone      7118 non-null   object 
 8   telefone.varias_linhas         7118 non-null   object 
 9   internet.servico_internet      7118 non-null   object 
 10  internet.seguranca_online      7118 non-null   object 
 11  internet.backup_online         7118 non-null   object 
 12  internet.protecao_dispositivo  7118 non-null   object

Index: 7118 entries, 0 to 7343 $\rightarrow$ tem índice sendo saltado, (os índices de Churn vazio).

In [76]:
# Resetando o index 0-7117
dados_sem_vazio.reset_index(drop=True, inplace=True)

In [77]:
dados_sem_vazio

id_cliente Churn cliente.genero  cliente.idoso cliente.parceiro  \
0     0002-ORFBO   nao       feminino              0              sim   
1     0003-MKNFE   nao      masculino              0              nao   
2     0004-TLHLJ   sim      masculino              0              nao   
3     0011-IGKFF   sim      masculino              1              sim   
4     0013-EXCHZ   sim       feminino              1              sim   
...          ...   ...            ...            ...              ...   
7113  5792-JALQC   nao       feminino              1              nao   
7114  5172-RKOCB   nao      masculino              0              sim   
7115  1934-MKPXS   nao      masculino              0              sim   
7116  5959-BELXA   sim      masculino              1              nao   
7117  4452-QIIEB   nao      masculino              0              nao   

     cliente.dependentes  cliente.tempo_servico telefone.servico_telefone  \
0                    sim                    9.0                       sim   
1                    nao                    9.0                       sim   
2                    nao                    4.0                       sim   
3                    nao                   13.0                       sim   
4                    nao                    3.0                       sim   
...                  ...                    ...                       ...   
7113                 nao                   52.0                       sim   
7114                 nao                   72.0                       sim   
7115                 sim                   33.0                       sim   
7116                 nao                   32.0                       sim   
7117                 nao                    7.0                       sim   

     telefone.varias_linhas internet.servico_internet  ...  \
0                       nao                       DSL  ...   
1                       sim                       DSL  ...   
2                       nao               fibra otica  ...   
3                       nao               fibra otica  ...   
4                       nao               fibra otica  ...   
...                     ...                       ...  ...   
7113                    sim                       DSL  ...   
7114                    sim               fibra otica  ...   
7115                    nao                       nao  ...   
7116                    sim               fibra otica  ...   
7117                    nao                       DSL  ...   

       internet.backup_online internet.protecao_dispositivo  \
0                         sim                           nao   
1                         nao                           nao   
2                         nao                           sim   
3                         sim                           sim   
4                         nao                           nao   
...                       ...                           ...   
7113                      nao                           sim   
7114                      sim                           nao   
7115  sem servico de internet       sem servico de internet   
7116                      nao                           nao   
7117                      nao                           nao   

     internet.suporte_tecnico    internet.tv_streaming  \
0                         sim                      sim   
1                         nao                      nao   
2                         nao                      nao   
3                         nao                      sim   
4                         sim                      sim   
...                       ...                      ...   
7113                      nao                      nao   
7114                      sim                      sim   
7115  sem servico de internet  sem servico de internet   
7116                      nao                      sim   
7117                      nao                      nao   

     

<h2>
<font color="#9de482"> 03 - Utilizando dados duplicados e nulos </font>
</h2>